otter/integration/lib/test_autoscale.py

```diff
@@ -98,3 +100,83 @@ def test_poll_until_timeout(self):
             self.assertNoResult(d)
         self.clock.advance(10)
         self.failureResultOf(d, TimedOutError)
+
+
+class GetServicenetIPs(SynchronousTestCase):
+    """
+    Tests for :func:`ScalingGroup.get_servicenet_ips`.
+    """
+    def setUp(self):
+        """Create our simulated clock and scaling group."""
+        self.clock = Clock()
+        self.pool = object()
+        self.queried_server_ids = []
+
+        class FakeRCS(object):
+            endpoints = {'nova': 'novaurl'}
+            token = "token"
+
+        @attributes(['code', 'fake_json_response'])
+        class Resp(object):
+            pass
+
+        class FakeTreq(object):
+
+            @classmethod
+            def get(cls, url, headers, pool):
+                self.assertIs(self.pool, pool)
+                self.assertEqual(["token"], headers.get('x-auth-token'))
+                url_segments = url.split('/')
+                self.assertEqual(3, len(url_segments))
+                self.assertEqual(['novaurl', 'servers'], url_segments[:2])
+                self.queried_server_ids.append(url_segments[-1])
+                return defer.succeed(Resp(code=200, fake_json_response={
+                    'server': {
+                        'addresses': {
+                            'private': [
+                                {'addr': '10.0.0.{0}'.format(
+                                    len(self.queried_server_ids)),
+                                 'version': 4}
+                            ]
+                        }
+                    }
+                }))
+
+            @classmethod
+            def json_content(cls, resp):
+                return defer.succeed(resp.fake_json_response)
+
+        self.rcs = FakeRCS()
+        self.sg = ScalingGroup(group_config={}, pool=self.pool,
+                               reactor=self.clock, treq=FakeTreq)
+
+    def test_queries_for_provided_server_ids(self):
+        """
+        If server IDs are provided, IPs are queried for those server IDs.
+        And if the same server ID is given multiple times, only one query is
+        made for any given server ID.
+        """
+        server_ids = ['1', '2', '2', '2', '3']
+        d = self.sg.get_servicenet_ips(self.rcs, server_ids)
+        result = self.successResultOf(d)
+        self.assertEqual(['10.0.0.1', '10.0.0.2', '10.0.0.3'],
```

Comment:
Each time that method is called, the length changes.  So the first time it's called, it appends to the list, and the length is 1.  The second time it's called, it appends to the list, and the length is 2.  Third time it's called, it appends to the list, and the length is 3.

At the end, definitely the length is 3, but since `len(self.queried_server_ids)` is called each time the function is called to generate a response, we get 3 different responses :)